### PR TITLE
Improved error message if insufficient permission to read a file

### LINKED ISSUE #27823: Misleading error messages when permissions denied to files
Please Post on the [relevant Forum post](https://forum.mantidproject.org/t/loadmask-algorithm-execution-error/635) when this is resolved.
Also reported seperately by Pascal.
### Expected behavior
If a file which needs to be accessed by Mantid, has its permissions denied, then the Error Message displayed to the user should say something along the lines of "Permissions Denied for File:..."

### Actual behavior
This was found for a script which ran the `LoadMask` algorithm and said that the algorithm failed:
The error is: Error in execution of algorithm LoadMask: SAXParse Exception: no element found in line 1 column 0. Unable to parse File: in “/user/local/mprogs/InstrumentFiles/merlin/mask_Dec2019.xml”
### Steps to reproduce the behavior
Not sure of a good test case as the related file has had its permissions corrected, but essentially run `LoadEmptyInstrument` for the relevant instrument and then `LoadMask` on a mask file with its permissions denied.

### Platforms affected
All

### PR DESCRIPTION
**Pull request by StephenSmith25**
**Description of work.**
This PR improves the error messages generated when the user has insufficient permissions to read a file. This is handled by attempting to open the file in the `FileValidator` class and if unsuccessful writing a error based on the system error number. The test for this has only be added for Linux and Mac where the `filesystem` boost header allows read permissions to easily be removed from a file.

**To test:**
1. On windows you can remove permissions to a file using a powershell script
2. Open notepad and save the following as permissions.ps1

```
param(
[string]$filename,
[string]$permissions
)

icacls.exe $filename /reset
icacls.exe $filename /GRANT:R "$($env:USERNAME):($permissions)"
icacls.exe $filename /inheritance:r
```

3. For the test file you'll need the mantid usage data. Locate `HYS_mask.xml` and save a copy as something like HYS_mask_locked.xml. 
4. Move the powershell script from above into the folder 
5. Open a powershell and do:
` .\permissions.ps1 HYS_mask_locked.xml D`
6. Now open workbench add the above folder to the path and run the script:
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

mask = LoadMask('HYS', 'HYS_mask_locked.xml')
```
You should get a warning about permissions denied.

**On Linux you can use chmod to remove read permissions and then perform the steps above.**

Fixes #27823 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by clayton-tom — DISMISSED** [review]
The error message now reveals what is actually going wrong and the code looks good 👍

File: Framework/Kernel/src/FileValidator.cpp [hunk 1]
```diff
@@ -90,6 +91,16 @@ std::string FileValidator::checkValidity(const std::string &value) const {
     return "File \"" + abspath + "\" not found";
   }
 
+  if (m_testExist && (Poco::File(value).exists())) {
+    std::ifstream in;
+    in.open(value.c_str());
+    if (!in) {
+      auto e = std::system_error(errno, std::generic_category(),
```
Comment (martyngigg) [inline Framework/Kernel/src/FileValidator.cpp:98]: Are we using `system_error` just to format the error message?

Comment (StephenSmith25) [reply Framework/Kernel/src/FileValidator.cpp:98]: Maybe there's a better way to do this, but I was using it to parse the system errror number into some error message that I could return

Comment (StephenSmith25) [reply Framework/Kernel/src/FileValidator.cpp:98]: Seems I can probably use `strerror` for this instead.. Ill change it

**Review by martyngigg — CHANGES_REQUESTED** [review]
Just a small code cleanup suggestion.

**Review by martyngigg — APPROVED** [review]

**Pull request merged**